Initial support for PoseAssets

RedHaze opened the pull request:
**Summary**

This is initial support for converting Unreal Engine Pose Assets to Blender shape keys. It integrates fairly seamlessly with the existing Blender plugin. The motivation here is that I suck at making shape keys by hand, so I did the logical thing which was spend more time than the initial manual process to automate it 😄.

Unfortunately, there are still some characters which appear to be using some other method to drive facial flexes (i.e. Hulk), so this won't cover everything.

I've added some rather hacky support for facial flexes to influence the mesh found in `FaceAcc` / `CustomCharacterFaceData` (?). Any tips on how to improve this aspect of the code would be greatly appreciated. Alternatively, I am OK with contributors directlying comitting to this PR.

If you'd like to contact me directly, please ping me in the Discord @RedHaze.

**ToDo**

- [x] Maybe a nicer way to know when to propagate pose data? Is this something available in Anim BP?
- [x] Pruning blend shapes that have no influence on vertex groups according to pose bones.
- [x] ~~Look into Fish Thicc's use of `CurveData` and why it's triggering warnings.~~ will be done later if there's demand
- [x] ~~Implement (or remove?) `CurveData` processing in the Blender plugin based on findings.~~ will be done later if there's demand
- [ ] Strange issues with eye lids (`R_upperLidDown_pose` vs `L_upperLidDown_pose` where R is clearly wrong) when importing Fish Thicc
- [ ] Grriz has pose data, but there's a bug of some sort preventing proper import (possibly because it is essentially two characters in one)

RedHaze committed:
Initial support for PoseAssets

h4lfheart committed:
apply facial anim curves

h4lfheart committed:
scuffed metahuman support (idle poses only bc no bone mappings yet)

h4lfheart committed:
adjust code style

RedHaze commented on FortnitePorting/Plugins/Blender/import_task.py:
```diff
@@ -563,14 +591,153 @@ def get_meta(search_props):
+                                if not pose_bone:
+                                    # For some reason, some of the bones start capitalized and some dont
+                                    bone_name = bone_name[0].capitalize() + bone_name[1:]
+                                    pose_bone = armature.pose.bones.get(bone_name)
```
I'm not sure if this is robust with respect to case sensitivity differences for bone names in Pose Asset data vs the armature. For example, when exporting Rufus, the Pose Asset refers to the bone name `l_brow_outer`, however, the armature has the bone named as `L_brow_outer`. Is there bone name normalization going on at some point during the export process? If so, could we stop doing that? Alternatively, I suppose we could create a mapping of lower-case bone names to `Bone` instances so that we can match `bone['Name'].lower()` here against those consistently. Thoughts?

h4lfheart replied:
shouldnt be any changes the bone names on export, easiest fix i think is just calling .casefold() and searching thru the existing bone list would be fine, could look at ue source to see exactly wht they doing tho

RedHaze replied:
Got this done in latest commit, please take a look when you have the chance and resolve this comment.

RedHaze commented on FortnitePorting/Plugins/Blender/import_task.py:
```diff
@@ -563,14 +591,153 @@ def get_meta(search_props):
+
+                        # Find the ARMATURE modifier
+                        armature_modifiers = [mod for mod in imported_mesh.modifiers if mod.type == "ARMATURE"]
+                        assert len(armature_modifiers) == 1, "expected at least one armature modifier"
```
How do we feel about my use of assertions? Should I change these to formatted exceptions?

h4lfheart replied:
tbh i've never really used assertions in python so i cant really comment on that
it might be a better design choice to let the export carry out without pose stuff if it doesnt work? bc right now there's no way to toggle pose asset loading so it could break entire assets which would be annoying for people who need the skin but arent using em

RedHaze replied:
Added some slightly better error handling w.r.t resetting pose bones on failed pose asset import. There may be partially incomplete keys, but otherwise I think it's an improvement. Let me know if you'd like to see any additional changes here.

RedHaze commented on FortnitePorting/Plugins/Blender/import_task.py:
```diff
@@ -563,14 +591,153 @@ def get_meta(search_props):
+                            if scale != Vector((1, 1, 1)):
+                                if entry['BoneName'] == 'faceAttach':
+                                    face_attach_scale = scale
+                                Log.warn(f"Non-zero scale: {entry}")
```
Exporting Ranger, I notice the warning appearing for bones other than `faceAttach`:
```
[FNPORTING] Non-zero scale: {'BoneName': 'R_eye_lid_upper_mid', ... 'Scale': {'X': 0.999998, 'Y': 1.0000657, 'Z': 0.9999363}}
[FNPORTING] Non-zero scale: {'BoneName': 'R_eye_lid_lower_mid', ..., 'Scale': {'X': 0.999998, 'Y': 1.0000249, 'Z': 0.9999771}}
```

They're close enough to `(1, 1, 1)` that it _probably_ wont make a difference, however, I wonder if other characters have weird scaling in their reference pose which also needs to be accounted for when creating location data.

Just want to reiterate that the scaled location is based on observation from an old export of facial data I have, so this _may_ be inaccurate / unnecessary. You can test this visually with Polar Patroller who has a `0.890` scaled `faceAttach` bone. I guess we can leave it in for now if there are no objections though?

h4lfheart replied:
i mean if it works it works, until we encounter a breaking issue i think it's fine

h4lfheart committed:
update cue4parse ref

RedHaze commented on FortnitePorting/Plugins/Blender/import_task.py:
```diff
@@ -563,14 +591,153 @@ def get_meta(search_props):
+                                edit_bone = pose_bone.bone
+                                post_quat = Quaternion(post_quat) if (post_quat := edit_bone.get("post_quat")) else Quaternion()
+
+                                assert pose_bone.rotation_quaternion == Quaternion(), "initial pose bone is not identity"
```
I think I'll just call `pose_bone.matrix_basis.identity()` and get rid of these assertions

h4lfheart replied:
bet.

RedHaze commented on FortnitePorting/Plugins/Blender/import_task.py:
```diff
@@ -563,14 +591,153 @@ def get_meta(search_props):
+                                assert pose_bone.rotation_quaternion == Quaternion(), "initial pose bone is not identity"
+                                q = post_quat.copy()
+                                if edit_bone.parent is None:
+                                    q.rotate(make_quat(rotation).conjugated())
```
This might be pointless since I imagine every face bone is always hooked up to `faceAttach`. Thoughts on removing this? Or perhaps we have an assertion / warning when this case appears.

h4lfheart replied:
yeah probably just remove it, only valid for the root bone which is never used

RedHaze commented on FortnitePorting/Plugins/Blender/import_task.py:
```diff
@@ -563,14 +591,153 @@ def get_meta(search_props):
                 meta = get_meta(["SkinColor"])
-            case "Head":
-                meta = get_meta(["MorphNames", "HatType", "PoseData"])
+            case "Head" | "Face":
```
We'll actually want to apply Pose Asset data to body (maybe just all armatures) as there is at least one case where the jaw clips into clothing otherwise (i.e. Cube Queen).

h4lfheart replied:
agreed, we should probably add important pose data stuff to every meta type and carry out the pose asset stuff afterwards no matter the type

RedHaze replied:
Done, let me know what you think.

RedHaze reviewed (commented):
Comments to @halfuwu and myself. I have some small code changes I'd like to add too, but larger suggestions may live in a future PR so as to not block this one.

RedHaze committed:
Ensure pose bone is identity before processing it

* Remove need for identtiy check assertions
* Simplify armature search. Blender may append `.001` or similar to object
  names, but that would be in addition to the original imported mesh name.
  So do a simple `in` check instead of splitting at `.`.

RedHaze committed:
Remove unnecessary code

RedHaze committed:
More Python code cleanup

h4lfheart committed:
push to v2.0.11 and simplify armature "finding" (lol)

RedHaze committed:
Restore original pose transforms on error or success

RedHaze committed:
Case insensitive bone name search

RedHaze committed:
Apply pose data to all character parts

RedHaze committed:
Forgot a small change

RedHaze committed:
Prevent creation of blank shape keys

RedHaze committed:
Update CUE4Parse commit

* Fix build error in FN

RedHaze committed:
Remove extra warnings / code pertaining to unfinished CurveData handling

h4lfheart commented on FortnitePorting/Plugins/Blender/import_task.py:
```diff
@@ -574,6 +600,128 @@ def get_meta(search_props):
+                for entry in reference_pose:
+                    scale = make_vector(entry.get('Scale'))
+                    if scale != Vector((1, 1, 1)):
+                        if entry.get('BoneName', '').casefold() == 'faceattach':
```
the bone name "FACIAL_C_FacialRoot" is used for metahuman face skins, does this need to be added here too?

h4lfheart replied:
metahuman skins are newer skins like lorenzo, rian, the ageless, etc.

h4lfheart commented on FortnitePorting/Plugins/Blender/import_task.py:
```diff
@@ -574,6 +600,128 @@ def get_meta(search_props):
+                        pose_bone.matrix_basis.identity()
+
+                        rotation = bone.get('Rotation')
+                        assert rotation.get('IsNormalized'), "non-normalized rotation unsupported"
```
cant we just normalize the rotations during the export process if they're not normalized

RedHaze replied:
Done.

h4lfheart commented on FortnitePorting/Export/ExporterInstance.cs:
```diff
@@ -51,6 +51,95 @@ public ExporterInstance(EExportTargetType exportType)
+
+        if (poseAsset.RetargetSourceAssetReferencePose is null) return;
+        if (poseAsset.Skeleton is null) return;
+        if (!poseAsset.Skeleton.TryLoad<USkeleton>(out var skeleton)) return;
```
does skipping the code below actually cause issues for export? (metahuman skins don't go through this at all)

h4lfheart reviewed (changes requested):
just some final comments for stuff that may be breaking/unnecessary, lmk whatcha think then we can merge!

RedHaze committed:
Normalize rotations during export

h4lfheart committed:
update to latest
